feat(fields): add select property to upload and relationship field

milamer opened the pull request:
## Description

This PR adds a new `select` property to the `Upload` and `Relationship` fields. This allows you to specify which fields you want to return from the related model. This is useful for reducing the amount of data that is returned from the server.

Related discussion: https://github.com/payloadcms/payload/discussions/1042 from @christian-reichart

As pointed out in the discussion the `afterRead` hook cannot be used to achieve this as the `afterRead` hook is called before the relationship is resolved. This means that the `afterRead` hook cannot be used to select certain fields of the related collection.

When using the `select` property you can specify which fields you want to return from the related model as an array of strings. For example, if you have a `Post` model with a `title` and `body` field, you can specify that you only want to return the `title` field when querying for a `Blog`:

```js
const Post = {
  slug: 'post',
  fields: [
    {
      name: 'title',
      type: 'text',
      required: true,
    },
    {
      name: 'body',
      type: 'richText',
      required: true,
    },
  ],
};

const Blog = {
  slug: 'blog',
  fields: [
    {
      name: 'title',
      type: 'text',
      required: true,
    },
    {
      name: 'post',
      type: 'relationship',
      relationTo: 'post',
      select: ['title'],
    },
  ],
};
```

The `id` field is always returned, so you don't need to specify it in the `select` property.




I also thought about making the `select` property a function that receives the document and the collection as arguments. And returns the fields that should be selected. This would allow you to select fields based on the document or collection. But I decided against this as it would make the `select` property more complex and I think it would be a rare use case. It would have looked something like this:


```js
select: (document, collection) => {
  if (document.isAdmin) {
    return collection.fields.map(field => field.name);
  }

  return ['id', 'name'];
}
```

Another option would be to use an `afterPopulate` hook. This would be similar to a normal `afterRead` hook but it would be called after the relationship is resolved. This would allow you to select fields based on the related document. But I think this would be a rare use case as well. And it would make the `select` property more complex.

What do you think?



- [x] I have read and understand the CONTRIBUTING.md document in this repository

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

I did not add tests for the RichText field.

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Existing test suite passes locally with my changes
- [x] I have made corresponding changes to the documentation

DanRibbens reviewed (commented):
At first glance this looks great! Have you tested it with field level access control and localization?

I like the simplicity. I wouldn't want to introduce `afterPopulate` as another hook if it can be avoided. I do like the idea of allowing `select` functions like you suggested but this could also just be added later at any time without a breaking change.

DanRibbens commented on src/fields/hooks/afterRead/relationshipPopulationPromise.ts:
```diff
@@ -53,6 +53,16 @@ const populate = async ({
+      const newRelationShipValue = {
+        id: relationshipValue.id,
+      };
+      field.select.forEach((fieldName) => {
```
Do we want to support nesting fields?

`select: ['group.someField`]` for example. We could split the fieldName on the "." and build up the objects accordingly.

milamer replied:
I also thought about adding a way to select nested fields, but we would have to decide how to select arrays.
Either we would have to select the index of the array, or we would have to select the field name of the value in the array and then select the whole array.

Example of the first option:

```js
const relationshipField = {
  name: 'purchase',
  type: 'relationship',
  relationTo: ['products', 'services'],
  select: ['name', 'price', 'images.0.url'],
};
// returns
{
  id: '123',
  name: 'Product Name',
  price: 100,
  images: [
    {
      id: '456',
      url: 'https://example.com/image.jpg'
    }
  ]
}
```

Example of the second option:

```js
const relationshipField = {
  name: 'purchase',
  type: 'relationship',
  relationTo: ['products', 'services'],
  select: ['name', 'price', 'images.url'],
};

// returns
{
  id: '123',
  name: 'Product Name',
  price: 100,
  images: [
    {
      id: '456',
      url: 'https://example.com/image.jpg'
    },
    {
      id: '789',
      url: 'https://example.com/image2.jpg'
    }
  ]
}
```

milamer commented:
> At first glance this looks great! Have you tested it with field level access control and localization?

Glad you like it. 

I didn't test it yet, will try to do that tomorrow and report back!

> I like the simplicity. I wouldn't want to introduce `afterPopulate` as another hook if it can be avoided. I do like the idea of allowing `select` functions like you suggested but this could also just be added later at any time without a breaking change.

Yeah, I also thought about it more and I the idea with the `select` function really grew on me. I think it's a great idea and I will implement it like that. So the select field can either be an array of strings or a function that returns an array of strings.

DanRibbens commented:
There are a couple of more issues that we came up with around this PR. 

1. The admin UI will break if the `useAsTitle` field for the relationship is not included in the fields coming back.
2. GraphQL schema will not to limit the fields available based on `select`. Also how will this work with the function based `select`?
3. Same for typescript generation as #2. Possibly the function will just be a Partial<RelationType>
4. There is going to be trouble with the `relationTo` having many different types support. This one will take some planning. The best way would probably be to key the object with the relationship slug for the fields needed or treat them all the same but omit fields that the collection doesn't have. Does this make sense? This feels quite complicated TBH.

With all these complexities I have to ask, is there a reason you wouldn't prefer to use GraphQL to be specific about the data you need instead of changing the REST API?

I want to be sure we're not introducing more overhead to maintain than we are getting value. How do you feel about all this?

milamer commented:
Thanks for the feedback!

> The admin UI will break if the useAsTitle field for the relationship is not included in the fields coming back.

But in this case this shouldn't be a problem, because the `useAsTitle` just uses top level, non relationship fields.
This means the select property should be able to be used without any problems.


IMO there is a bigger issue with types here. The `useAsTitle` is just a string, but it should be a qualified field name.
I wanted to give the type system a try, but didn't have the time yet. I'll try to do it soon.

> GraphQL schema will not to limit the fields available based on select. Also how will this work with the function based select?

I added a note to the docs that the `select` property will not work with GraphQL.
I am not sure if it is possible to implement it there.
Do you have any ideas?

> Same for typescript generation as `#2.` Possibly the function will just be a Partial

I already started to work on this. Ideally the generated types should only have
the fields that are returned by the `select` property. `Partial<>` would be a good start.

> There is going to be trouble with the relationTo having many different types support. This one will take some planning. The best way would probably be to key the object with the relationship slug for the fields needed or treat them all the same but omit fields that the collection doesn't have. Does this make sense? This feels quite complicated TBH.

I don't quite follow. Can you perhaps give an example?
Currently fields that return `undefined` are ignored, so I think it should work fine.

This is the current type for the `select` property:

```ts
type KeysOfUnion<T> = T extends T ? keyof T : never;

export type SelectFunction<T extends TypeWithID> = (
  options: SelectOptionsProps<T>
) => Array<Exclude<KeysOfUnion<T>, "id">> | true;

export type SelectOptions<T extends TypeWithID> =
  | Array<Exclude<keyof T, "id">>
  | SelectFunction<T>;
```

This means only fields that are defined on all the related collection are allowed.
Should this be changed to allow fields that are only defined on some of the related collections?

Currently you have to use the function to achieve exactly that. It is possible to return an array of fields that are only defined on some of the related collections.
But the type system isn't 100% correct here, because you don't have to narrow the type down to use only the fields of the current collection.

The ideal way would be to have a type that is the union of all the related collections and then use that to narrow down the type. Also the data Type should be generated based on the related collections.

Example:

```ts
const func: SelectFunction<Collection1 | Collection2 | Collection3> = ({
  data,
  collection,
}) => {
  if (collection.slug === "collection1") {
    return ["a", "b"];
  }
  return ["a"];
};
```

> With all these complexities I have to ask, is there a reason you wouldn't prefer to use GraphQL to be specific about the data you need instead of changing the REST API?

Href links to related collection is a core concept for CMSs. It is also used in rich text fields. Always using GraphQL for this would be a huge overhead.
And then I don't see why REST is supported at all.

I think payload does a great job of providing the data of the whole page in one request.
This is a huge advantage over other CMSs where you have to know the depth or make multiple requests to get the data you need.

IMO the benefit of GraphQL is that you can query the data you need. 
But this is not really useful for a classic website with pagebuilder like nested blocks of content. THE CMS is kept in sync with the website, so you always know what data you need.
If we would use an app that has multiple versions and needs to deprecate older API versions, then it would be a different story.

I'm also working to support the select function in rich text fields. This will function the same way as the relationship field.

DanRibbens commented:
Great discussion, everything you say makes sense.

For arrays and block nesting, go with this option IMO:
> `select: ['name', 'price', 'images.url'],` 

Instead of `images.0.url`. We don't do limiting on iterables anywhere in Payload that I'm aware of besides pagination at the top level and this seems to be okay.

> > The admin UI will break if the useAsTitle field for the relationship is not included in the fields coming back.
> 
> But in this case this shouldn't be a problem, because the useAsTitle just uses top level, non relationship fields.
> This means the select property should be able to be used without any problems.

You're right, the relationships get queried directly by the admin UI so this is a non-issue.

> > GraphQL schema will not to limit the fields available based on select. Also how will this work with the function based select?
>
> I added a note to the docs that the select property will not work with GraphQL.
> I am not sure if it is possible to implement it there.
> Do you have any ideas?

We just need to add a condition to the `if` on this line https://github.com/payloadcms/payload/pull/1288/files#diff-d08122d526f09ae6bb53de0f8ad0b11358d1b20932e8577633778e0ef06e6176R56 to include `&& req.payloadAPI !== 'GraphQL'. That way GraphQL queries can still continue to select granularly what is needed and there is no need for any schema changes.

> > There is going to be trouble with the relationTo having many different types support. This one will take some planning. The best way would probably be to key the object with the relationship slug for the fields needed or treat them all the same but omit fields that the collection doesn't have. Does this make sense? This feels quite complicated TBH.
>
> I don't quite follow. Can you perhaps give an example?
> Currently fields that return undefined are ignored, so I think it should work fine.

Forget this too. I was concerned about not having granular control. Suppose your field had `relationTo: ['videos', 'images']` and both relationships have a field for `size` but you only need `size` for images and not video. `Select` would have to support `{ videos: ['length'], images: ['size'] }`, but I think it is okay to just flatten it as: `['length', 'size']`. This edge case only would exist when somebody has the same field names across relationshis and only wants to select in one relationship and not another.

> Should this be changed to allow fields that are only defined on some of the related collections?
I think what you have is okay as a start, but improving it would be great if we can. Could we use the Pick utility type here to narrow the fields to the `select`?
https://www.typescriptlang.org/docs/handbook/utility-types.html#picktype-keys

Let me know if I missed any questions and thank you for the deep dive into the work here!

milamer commented:
Thanks for the great feedback. I will try to address all of your points.

> select: ['name', 'price', 'images.url'],

In the first implementation I would skip nested fields as this could have unexpected results. And would make the type system more complex.

> You're right, the relationships get queried directly by the admin UI so this is a non-issue.

We checked that again and not including the `useAsTitle` field in the query results in the admin UI not showing the title of the related collection, but the ID instead.

> We just need to add a condition to the if on this line https://github.com/payloadcms/payload/pull/1288/files#diff-d08122d526f09ae6bb53de0f8ad0b11358d1b20932e8577633778e0ef06e6176R56 to include `&& req.payloadAPI !== 'GraphQL'. That way GraphQL queries can still continue to select granularly what is needed and there is no need for any schema changes.

Ah thanks for that. I added the condition.

> This edge case only would exist when somebody has the same field names across relationshis and only wants to select in one relationship and not another.

This should be possible with the select function :D

> I think what you have is okay as a start, but improving it would be great if we can. Could we use the Pick utility type here to narrow the fields to the select?

I checked out [json-schema-to-typescript](https://github.com/bcherny/json-schema-to-typescript), but I don't think it's possible to use Generics to narrow the fields to the select.
I am not sure if it's possible to use the Pick utility type here.

Do you have any ideas?


I also added the support for the select function in richtext fields. This will function the same way as the relationship field, but will only support the `select` function.

Thanks again for the help!

christian-reichart commented on docs/fields/rich-text.mdx:
````diff
@@ -368,3 +369,37 @@ import type {
 
 ```
+
+### Selecting fields
````
Could be maybe just a reference to the relationship field section?

milamer replied:
No as I wanted to have a different API for richt-text and relationship/upload

christian-reichart commented on src/fields/config/schema.ts:
```diff
@@ -318,6 +326,7 @@ export const blocks = baseField.keys({
 export const richText = baseField.keys({
   type: joi.string().valid('richText').required(),
   name: joi.string().required(),
+  select: joi.func(),
```
This should allow array aswell for a simple usecase.
() => { return ['title'] } works and is basically the same

milamer replied:
That is true, but I think it should be encouraged to use a function and return true as default

```ts
({collection}) => {
  if (isMyCollection(collection) return ['title']
  return true
}
```

Perhaps it should be better documented that this is the desired approach as relationships can be added to the richText without the richText knowing of it.

milamer commented:
Hi @DanRibbens,

I was wondering if you have a moment to look over this pull request. I know you're probably busy, but I would really appreciate your feedback. I've made some changes and I think they would be a great addition to the project.

Let me know if you have any questions or if there's anything I can help with.

Thanks for your time and for maintaining this project!

DanRibbens commented on src/fields/config/schema.ts:
```diff
@@ -273,6 +273,10 @@ export const upload = baseField.keys({
   name: joi.string().required(),
   maxDepth: joi.number(),
+  select: joi.alternatives().try(
+    joi.array(),
```
You can be more specific on the array:
```
select: joi.alternatives().try(
    joi.array().items(joi.string()),
    joi.func(),
  ),
```

DanRibbens commented on src/fields/config/schema.ts:
```diff
@@ -310,6 +314,10 @@ export const relationship = baseField.keys({
     joi.func(),
   ),
+  select: joi.alternatives().try(
+    joi.array(),
```
Same as above.

DanRibbens reviewed (commented):
Overall this is looking really good!
I know you said you will handle nested structures, I think that would really make this feature complete.

SiblingData should be good the way you are did it. I tested and it worked as I'd expect.

One additional thought, could we add the `req` to the select function argument object? That would allow somebody to write a `select` function to be dynamic based query params. I wouldn't build anything into the core for handling the URL since that gets us into GraphQL territory, but allowing this to be an option for projects and plugins could be really powerful.

milamer commented:
> Overall this is looking really good!
> I know you said you will handle nested structures, I think that would really make this feature complete.

Deep Picking nested fields is really complicated. Most of all the types. I wrote a "small" function that creates a deep pick and handles a few edges cases like arrays:
```ts
/* eslint-disable no-param-reassign */

// this is a placeholder for a complex type
type DeepPick<TObject, TPaths extends string> = Partial<TObject>;


export function deepPick<T>(
  obj: T,
  paths: Array<string>,
): DeepPick<T, typeof paths[number]> {
  const result: any = {};
  paths.forEach((path) => {
    const pathParts = path.split('.');
    deepPickTo(obj, pathParts, result);
  });
  return result;
}

function deepPickTo<T>(obj: T, path: Array<string>, result: any): boolean {
  if (typeof result !== 'object') {
    return false;
  }
  const pathPart = path[0];
  if (!(pathPart in obj)) {
    return false;
  }
  const value = obj[pathPart];
  if (path.length === 1) {
    result[pathPart] = obj[pathPart];
    return true;
  }
  if (Array.isArray(value)) {
    let someFound = false;
    const newArray = value.map((subObj) => {
      const subResult = result[pathPart] || {};
      if (deepPickTo(subObj, path.slice(1), subResult)) {
        someFound = true;
        return subResult;
      }
      return {};
    });

    if (someFound) {
      result[pathPart] = newArray;
      return true;
    }
    return false;
  }
  const subResult = result[pathPart] || {};
  if (deepPickTo(value, path.slice(1), subResult)) {
    result[pathPart] = subResult;
    return true;
  }
  return false;
}

```

Examples:
```ts
const obj = {
  a: {
    b: {
      c: 1,
      d: 2,
    },
    e: 3,
  },
};
console.log(deepPick(obj, ['a.b.c', 'a.e']));
// { a: { b: { c: 1 }, e: 3 } }

const obj2 = {
  a: [
    {
      b: {
        c: 1,
        d: 2,
      },
    },
    {
      b: {
        c: 3,
        d: 4,
      },
    },
  ],
};
console.log(deepPick(obj2, ['a.b.c']));
// { a: [ { b: { c: 1 } }, { b: { c: 3 } } ] }

const obj3 = {
  a: [
    {
      b: {
        c: 1,
        d: 2,
      },
    },
    {
      e: {
        f: 3,
        g: 4,
      },
    },
  ],
};
console.log(deepPick(obj3, ['a.b.c']));
// { a: [ { b: { c: 1 } }, {} ] }
```

Like I said in one of my previous comments I checked out [json-schema-to-typescript](https://github.com/bcherny/json-schema-to-typescript), but I don't think it's possible to use Generics to narrow the fields to the select.
I am not sure if it's possible to use the Pick utility type here. Furthermore if we use the deepPick function it would be even harder.

Do you  have any input for this? 

> SiblingData should be good the way you are did it. I tested and it worked as I'd expect.
> 
> One additional thought, could we add the req to the select function argument object? That would allow somebody to write a select function to be dynamic based query params. I wouldn't build anything into the core for handling the URL since that gets us into GraphQL territory, but allowing this to be an option for projects and plugins could be really powerful.

Cool idea! I added that to the function property object.

DanRibbens commented on docs/fields/relationship.mdx:
```diff
@@ -98,6 +99,54 @@ You can learn more about writing queries [here](/docs/queries/overview).
 
+### Selecting fields
+
+By default, the entire related document is returned when querying for a relationship field.
```
Oh my bad! I forgot about local, no need to change the condition.

milamer commented:
Hi @DanRibbens,

I hope this message finds you well. I just wanted to check in on the progress of the PR and see if there is anything else that needs to be done before it can be merged. We are really excited to use this feature in our navigation and would prefer to avoid using a fallback like GraphQL if possible. Any updates or guidance you can provide would be greatly appreciated.

I also wanted to take this opportunity to wish you and your family a happy holiday season. I hope you have a wonderful time with your loved ones and enjoy some rest and relaxation.

Thank you for your time and assistance.
Chris

christian-reichart commented on src/fields/hooks/afterRead/relationshipPopulationPromise.ts:
```diff
@@ -55,6 +56,19 @@ const populate = async ({
+      if (fieldsOrTrue !== true) {
+        // if fieldsOrTrue is true then we want to return the entire related document
+        relationshipValue = deepPick(relationshipValue, ['id', ...fieldsOrTrue]);
+      }
```
Can't comment on unchanged lines in Github, but line 118 should be changed to:

`if (req.locale === 'all' && typeof siblingDoc[field.name] === 'object' && siblingDoc[field.name] !== null) {`

since typeof 'object' will be true with null, leading to the Object.keys[] to fail in the next line.

JarrodMFlesch replied:
Yep. One of those awesome JS things, it hurts you once and you never forget.

milamer replied:
This was a different error and was already fixed

christian-reichart commented:
@JarrodMFlesch  @DanRibbens What's the state on this PR? We've currently patched this in because our application would not be usable without the stripping away of fields for internal linking.

Do you think there will be progress in the future and you just don't have the time right now? Or can you forsee problems with this PR leading to it being rejected? It would be important for us to know if we should keep our patch of find a patch independent solution to the problem.

Thanks for any insights, and please don't feel pressured, just wanted to know the state :)

DanRibbens reviewed (changes requested):
It looks good overall. We had some concerns with the (necessary) complexity of deepPick.ts. 

I'd like to look at the types as we have added the package `ts-essentials` which might have some parts we could import and re-use.

Then for us to support this in the future it would be a good idea to add a couple unit tests in a new file deepPick.spec.ts or increase coverage in the int.spec with assertions for nested field examples. As it stands I'm afraid to touch this code TBH.

We also need to get the conflicts resolved before merging.

I think we're almost there and if you need help with any of this LMK.

roydigerhund commented:
Any updated on this PR? I would need the feature in a new project and I would like to know if I should wait for the feature or switch to a third-party mongo client or GraphQL.

christian-reichart commented:
@matthijs166 Your second point, setting a maxDepth to a relationship field, is already possible: https://payloadcms.com/docs/fields/relationship

milamer commented:
Hey everyone,

Thanks for reaching out about the pull request. I'm swamped with a big work project right now, but I want you to know that I appreciate your requests and I still want to finish this pull request at soon as possible.

matthijs166 commented:
@milamer @christian-reichart How far is this PR? Is there any help needed?

Our application is live at the moment and requests are 2,5 MB large now. We got a lot of relationships active on the live site.
If this takes a while, I'm considering rebuilding it to the GraphQL implementation.

christian-reichart commented:
@matthijs166 we are using https://www.npmjs.com/package/patch-package currently in our version for the time being. PR make still take a while but I think @milamer is determined to finish it once work slows down a bit. ;)

roydigerhund commented:
> @milamer @christian-reichart How far is this PR? Is there any help needed?
> 
> Our application is live at the moment and requests are 2,5 MB large now. We got a lot of relationships active on the live site. If this takes a while, I'm considering rebuilding it to the GraphQL implementation.

I'm using the GraphQL API (might work with the payload API client too) and the beforeRead hook to filter out fields I don't wanna include for specific requests. This might be a solution for you as well.

matthijs166 commented:
> @matthijs166 we are using https://www.npmjs.com/package/patch-package currently in our version for the time being. PR make still take a while but I think @milamer is determined to finish it once work slows down a bit. ;)
@christian-reichart Thank you for the tip of the patch-package! Going to use it immediately.

> I'm using the GraphQL API (might work with the payload API client too) and the beforeRead hook to filter out fields I don't wanna include for specific requests. This could be a solution for you as well.
@roydigerhund Rewrite to GraphQL is for this project a big job. Implementing this feature would make live way easier.

roydigerhund commented:
@matthijs166 I just checked it, it works with the Payload client too and might be more useful in some situations since you can also disable Rich-Text fields on the first level. Those usually get parsed which adds more loading time. Here is my example where I removed the Rich-Text fields in specific cases:

```
    beforeRead: [
      async ({ doc, req, query }) => {
        // change the checks based on your query client and parameters
        // keep in mind, the admin page will be affected too
        const isFetchingAllArticles = query === undefined || Object.keys(query).length === 0;
        const isFetchingArticlesInCategory = query && query.category;
        const isFetchingLinkRelation = req.user === 'link-relation';
        if (isFetchingAllArticles || isFetchingArticlesInCategory || isFetchingLinkRelation) {
          return {
            ...doc,
            content: undefined,
            usefulHints: undefined,
          };
        }
        return doc;
      },
    ],
```

milamer commented:
Hey everyone,

I hope you're doing well. I wanted to provide you with an update on my PR. After taking some time to review it again, I have made several improvements to enhance the functionality and maintainability. Here's a summary of the changes:

- Unit tests: I have added comprehensive unit tests for the deepPick method. This ensures the reliability and correctness of the functionality.
- API enhancement: In order to get better typing I moved to a select object instead of a string. This is similar to the select API used by Prisma. I have  modified it to utilize the following syntax: { [key]: true | { ... } }. This leads to a better-typed responses from the DeepPick helper.
- Type cleanup: I improved the deepPick type to be more readable and easier to understand.

I run all tests locally and I believe this commit should be the final one.

@DanRibbens, please take another look at the PR and provide your feedback. Your input is greatly appreciated.

Thank you all for your patience.

Cheers

DanRibbens commented:
I'll look soon! I want to get this one merged and released. Really appreciate all the work you put in.

jmikrut commented:
I think it should remain as select for sure, because that’s what it’s called in SQL and the word populate is often used to describe populating related documents

ssyberg commented:
Apologies for not finding my answer in this thread but it's a long one! Does this provide `select` field configuration only on relationships? I'm assuming this is still not supported by the API?

jmikrut commented:
OK - here's an update on this PR.

First off, thank you everyone for your patience here. This is a very important feature to get right and I think the team really should have prioritized a deep-dive here long ago. We just took a minute to review this from top to bottom and have a few thoughts.

### Reducing data fetched

Obviously this is the big driver for this feature. But this PR only handles this halfway, and still fetches _all_ data from the DB, and then simply reduces it on the server. It would be great if we could extend this PR so that we use ORM features to select only specific fields when we go to retrieve the data itself from the DB (discussed in #2458).

If we can make sure that our established pattern simply _**allows for this**_ in the future, then the Payload team is OK with merging this PR in advance of actually reducing the data returned from the DB. But we want to make sure that this can be a fast-follow, and not introduce any breaking API changes when we do implement this on the DB level.

For example, to leverage this pattern on the DB, the fields would simply never be returned from the query, and never be present in the `afterRead` hooks. So there might be some introduced inconsistencies with the way that things are currently implemented if we end up adding DB-level selection in the future. Maybe not though - I just thought I'd double-check here.

### Need a way for this to work on top-level docs as well

This is not only important for related docs - it's also important for top-level docs. What is the pattern there, or what could it be in the future? Should that be a programmatic REST API / local API query parameter like `?select=title,description` or similar? Again, this could be a fast follow and should not prevent merging this, but we need to plan for it. Maybe we could just expose another local API property like `select` that works on top-level and the same could be present for REST API?

As long as we have a pattern defined, so that we can remain consistent, this should not hold us up from finishing / merging.

### "Magic" query params

As Dan mentioned above, we likely need some way to dynamically determine the fields that are selected for a given REST API query. Typically, this is done via driving the entire field selection from query params, like `?select=title,description` or similar. Like, the whole thing would operate through query params. But, that pattern could quickly become VERY cumbersome to use, because you have to really go crazy with that syntax reducing not only top-level document fields, but all relation documents at all levels. That would potentially be a lot of query variables.

So that said, the Payload team does indeed like the simplicity of having these `select` functions defined in code. And we envison that doing so could also allow for you to create your _own_ system of "magic query params" like, say, `?selectFields=blog` which you'd have access to in your `select` prop from the `req` arg, that knows exactly what YOU are trying to do. That way it's nice and clean from the query perspective. SO I think we're sold on the functional, config-based approach. Just need to make sure that we have a way to make this easy, and I think magic query params could be a move.

Should be completely possible with what you have now, but I just thought that I would note that we are now all onboard with this PR as a whole, and the above query params approach is how _we_ will probably use this functionality. Good stuff.

### Rich text relations

How are rich text relations handled in this PR? I have not jumped into code yet, so apologies if this is already solved, but I would like to see a system of something like this:

```ts
const myRichText: Field = {
  name: "myRichText",
  select: {
    pages: ({ req, doc, whatever }) => {
      return ["title", "slug"];
    },
    posts: ({ req, doc, whatever }) => {
      return ["slug"];
    },
  },
};
```

@milamer I think this is what you've suggested above, although I'd like to make sure.

### Handling many relationTos

There's a bit of conversation above on many `relationTo`s above, but the above rich text field pattern seems like it could work for relationship fields with many `relationTo`s as well. I'd like to make sure that we match the config approach here.

### TypeScript ramifications

Also, finally, yes, the TS magic here will be incredibly difficult to accomplish. I don't see how we will ever be able to dynamically and accurately compute the values of the selected fields if they are functional (or based on a magic `req.query` or anything similar.)

> Therefore, instead of trying to generate the types automatically, I have decided to export the DeepPick and DeepPickKeys types so that the user can create their own types using these as a starting point.

I think this makes sense. This also means that I don't think that we should let TS experience overrule a clean config experience.

> I have modified it to utilize the following syntax: { [key]: true | { ... } }. This leads to a better-typed responses from the DeepPick helper.

If we don't need the object shape for TS, I think returning an array of strings from the `select` function seems like the cleanest API possible. So I'm kinda thinking that we would return to that.

### Overall

@milamer I think we should work together on brainstorming the above issues, and then once we are comfortable with solutions on all aspects, I'd like if you were to update the original PR message above with exactly what the "usage" of this new feature would look like, with config examples. I am having a tough time following the thread at this point admittedly ☠️

We REALLY appreciate your help here, as this is a giant feature and is very important! We're also happy to take over as well from here because you've already done a lot. But we have a few other roadmap items to get to so it'll be a bit before we can take this and run with it.

denolfe commented:
Hey @milamer, this would still be a great addition to Payload. We've recently migrated our repository to a monorepo, which means in order for this PR to be viable, you'll need to update it to target the `main` branch and change the code to target `packages/payload`.

Let me know if this is something you can take on.
